docs/source/developers/reviewing.rst

```diff
@@ -0,0 +1,226 @@
+----------
+
+- Arrow is a set of open source libraries and will be used in a very wide
+  array of contexts (including fiddling with deliberately artificial data
+  at a Jupyter interpreter prompt).  If you are writing a public API, make
+  sure that it won't crash or produce undefined behaviour on unusual (but
+  valid) inputs.
+
+- When a non-trivial algorithm is implemented, defense-in-depth checks can
+  be useful to catch potential problems (such as debug-only assertions, if
+  the language allows them).
+
+- APIs ingesting potentially untrusted data, such as on-disk file formats,
+  should try to avoid crashing or produce silent bugs when invalid or
+  corrupt data is fed to them.  This can require a lot of care that is
+  out of the scope of regular code reviews (such as setting up fuzz testing
+  - XXX link to current fuzz setup), but basic checks can still be suggested
+  at the code review stage.
+
+- When calling foreign APIs, especially system functions or APIs dealing with
+  input / output, do check for errors and propagate them (if the language
+  does not propagate errors automatically, such as C++).
+
+Performance
+-----------
+
+- Think about performance, but do not obsess over it.  Algorithmic complexity
+  is important if input size may be "large" (the meaning of large depends
+  on the context: use your own expertise to decide!).  Micro-optimizations
+  improving performance by 20% or more on performance-sensitive functionality
+  may be useful as well; lesser micro-optimizations may not be worth the
+  time spent on them, especially if they lead to more complicated code.
+
+- If performance is important, measure it.  Do not satisfy yourself with
+  guesses and intuitions (which may be founded on incorrect assumptions
+  about the compiler or the hardware).
+
+- Try to avoid trying to trick the compiler/interpreter/runtime by writing
+  the code in a certain way, unless it's really important.  These tricks
+  are generally dependent on platform details that may become obsolete,
+  they are often brittle, and they can make code harder to maintain
+  (a common question that can block contributors is "what should I do
+  about this weird hack that my changes need to remove"?).
+
+- Avoiding worst-case degenerescence (such as memory blowups when a size
+  estimate is too imprecise) may be more critical than trying to improve
+  the common case by a small amount.
+
+Documentation
+-------------
+
+These guidelines should ideally apply to both prose documentation and
+in-code docstrings.
+
+- Look for ambiguous / sub-informative wording.  For example, "it is an error
+  if ..." is less informative than either "An error is raised if ... " or
+  "Behaviour is undefined if ..." (the first phrasing doesn't tell the
+  reader what actually *happens* on such an error).
+
+- Be mindful about spelling, grammar, expression, concision.  This may seem
+  like an obvious concern, but not all people are not accustomed to care about
+  this.
```

Comment:
Perhaps, instead of saying it's obvious but that not all people care about it, we could phrase it more positively, for example, saying that these things help smooth communication, or something like that?

Reply:
Yes, possibly something like "Clear communication in a group with a range of linguistic and programming backgrounds enables more efficient collaborative development and maintenance."

Reply:
Hmm, how something like:
> People have varying degrees of adherence to these concerns, but they contribute to a better documentation.

Reply:
What about if, instead of talking about when people don't do the thing we want them to do, we phrase it in terms of the result of them doing the thing we want them to do:

>  Clear communication is essential for effective collaboration with people from a wide range of backgrounds, and contributes to better documentation.

Reply:
Ok :-)